diff -Naur sundials-7.4.0/include/sunlinsol/sunlinsol_lapackband.h patched_sundials/include/sunlinsol/sunlinsol_lapackband.h
--- sundials-7.4.0/include/sunlinsol/sunlinsol_lapackband.h	2025-06-23 18:49:58
+++ patched_sundials/include/sunlinsol/sunlinsol_lapackband.h	2025-07-25 14:08:46
@@ -40,7 +40,7 @@
 struct _SUNLinearSolverContent_LapackBand
 {
   sunindextype N;
-  sunindextype* pivots;
+  int* pivots;
   sunindextype last_flag;
 };
 
diff -Naur sundials-7.4.0/include/sunlinsol/sunlinsol_lapackdense.h patched_sundials/include/sunlinsol/sunlinsol_lapackdense.h
--- sundials-7.4.0/include/sunlinsol/sunlinsol_lapackdense.h	2025-06-23 18:49:58
+++ patched_sundials/include/sunlinsol/sunlinsol_lapackdense.h	2025-07-25 14:11:40
@@ -40,7 +40,7 @@
 struct _SUNLinearSolverContent_LapackDense
 {
   sunindextype N;
-  sunindextype* pivots;
+  int* pivots;
   sunindextype last_flag;
 };

diff -Naur sundials-7.4.0/src/sundials/sundials_lapack_defs.h patched_sundials/src/sundials/sundials_lapack_defs.h
--- sundials-7.4.0/src/sundials/sundials_lapack_defs.h	1970-01-01 01:00:00
+++ patched_sundials/src/sundials/sundials_lapack_defs.h	2025-06-02 13:45:56
@@ -0,0 +1,103 @@
+/* -----------------------------------------------------------------
+ * Programmer: Cody J. Balos @ LLNL
+ * -----------------------------------------------------------------
+ * SUNDIALS Copyright Start
+ * Copyright (c) 2002-2023, Lawrence Livermore National Security
+ * and Southern Methodist University.
+ * All rights reserved.
+ *
+ * See the top-level LICENSE and NOTICE files for details.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ * -----------------------------------------------------------------*/
+
+#ifndef _SUNDIALS_LAPACK_H
+#define _SUNDIALS_LAPACK_H
+
+#include <sundials/sundials_types.h>
+
+#ifdef __cplusplus  /* wrapper to enable C++ usage */
+extern "C" {
+#endif
+
+/*
+ * ==================================================================
+ * Blas and Lapack functions
+ * ==================================================================
+ */
+
+#if defined(SUNDIALS_F77_FUNC)
+
+#define dgbtrf_f77      SUNDIALS_F77_FUNC(dgbtrf, DGBTRF)
+#define dgbtrs_f77      SUNDIALS_F77_FUNC(dgbtrs, DGBTRS)
+#define dgetrf_f77      SUNDIALS_F77_FUNC(dgetrf, DGETRF)
+#define dgetrs_f77      SUNDIALS_F77_FUNC(dgetrs, DGETRS)
+
+#define sgbtrf_f77      SUNDIALS_F77_FUNC(sgbtrf, SGBTRF)
+#define sgbtrs_f77      SUNDIALS_F77_FUNC(sgbtrs, SGBTRS)
+#define sgetrf_f77      SUNDIALS_F77_FUNC(sgetrf, SGETRF)
+#define sgetrs_f77      SUNDIALS_F77_FUNC(sgetrs, SGETRS)
+
+#else
+
+#define dgbtrf_f77      dgbtrf_
+#define dgbtrs_f77      dgbtrs_
+#define dgetrf_f77      dgetrf_
+#define dgetrs_f77      dgetrs_
+
+#define sgbtrf_f77      sgbtrf_
+#define sgbtrs_f77      sgbtrs_
+#define sgetrf_f77      sgetrf_
+#define sgetrs_f77      sgetrs_
+
+#endif
+
+/* LAPACK */
+
+extern void dgbtrf_f77(const int *m, const int *n,
+                       const int *kl, const int *ku,
+                       double *ab, int *ldab, int *ipiv,
+                       int *info);
+
+extern void dgbtrs_f77(const char *trans, const int *n,
+                       const int *kl, const int *ku,
+                       const int *nrhs, double *ab,
+                       const int *ldab, int *ipiv,
+                       double *b, const int *ldb, int *info);
+
+
+extern void dgetrf_f77(const int *m, const int *n, double *a,
+                       int *lda, int *ipiv,
+                       int *info);
+
+extern void dgetrs_f77(const char *trans, const int *n,
+                       const int *nrhs, double *a,
+                       const int *lda, int *ipiv, double *b,
+                       const int *ldb, int *info);
+
+extern void sgbtrf_f77(const int *m, const int *n,
+                       const int *kl, const int *ku,
+                       float *ab, int *ldab, int *ipiv,
+                       int *info);
+
+extern void sgbtrs_f77(const char *trans, const int *n,
+                       const int *kl, const int *ku,
+                       const int *nrhs, float *ab,
+                       const int *ldab, int *ipiv,
+                       float *b, const int *ldb, int *info);
+
+extern void sgetrf_f77(const int *m, const int *n, float *a,
+                       int *lda, int *ipiv,
+                       int *info);
+
+extern void sgetrs_f77(const char *trans, const int *n,
+                       const int *nrhs, float *a,
+                       const int *lda, int *ipiv,
+                       float *b, const int *ldb, int *info);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif
diff -Naur sundials-7.4.0/src/sunlinsol/lapackband/sunlinsol_lapackband.c patched_sundials/src/sunlinsol/lapackband/sunlinsol_lapackband.c
--- sundials-7.4.0/src/sunlinsol/lapackband/sunlinsol_lapackband.c	2025-06-23 18:49:58
+++ patched_sundials/src/sunlinsol/lapackband/sunlinsol_lapackband.c	2025-07-25 14:09:44
@@ -112,7 +112,7 @@
   content->pivots    = NULL;
 
   /* Allocate content */
-  content->pivots = (sunindextype*)malloc(MatrixRows * sizeof(sunindextype));
+  content->pivots = (int*)malloc(MatrixRows * sizeof(int));
   if (content->pivots == NULL)
   {
     SUNLinSolFree(S);
@@ -148,7 +148,7 @@
 
 int SUNLinSolSetup_LapackBand(SUNLinearSolver S, SUNMatrix A)
 {
-  sunindextype n, ml, mu, ldim, ier;
+  int n, ml, mu, ldim, ier;
 
   /* check for valid inputs */
   if ((A == NULL) || (S == NULL)) { return SUN_ERR_ARG_CORRUPT; }
@@ -177,7 +177,7 @@
 int SUNLinSolSolve_LapackBand(SUNLinearSolver S, SUNMatrix A, N_Vector x,
                               N_Vector b, SUNDIALS_MAYBE_UNUSED sunrealtype tol)
 {
-  sunindextype n, ml, mu, ldim, one, ier;
+  int n, ml, mu, ldim, one, ier;
   sunrealtype* xdata;
 
   /* check for valid inputs */
diff -Naur sundials-7.4.0/src/sunlinsol/lapackdense/sunlinsol_lapackdense.c patched_sundials/src/sunlinsol/lapackdense/sunlinsol_lapackdense.c
--- sundials-7.4.0/src/sunlinsol/lapackdense/sunlinsol_lapackdense.c	2025-06-23 18:49:58
+++ patched_sundials/src/sunlinsol/lapackdense/sunlinsol_lapackdense.c	2025-07-25 14:10:33
@@ -113,7 +113,7 @@
   content->pivots    = NULL;
 
   /* Allocate content */
-  content->pivots = (sunindextype*)malloc(MatrixRows * sizeof(sunindextype));
+  content->pivots = (int*)malloc(MatrixRows * sizeof(int));
   if (content->pivots == NULL)
   {
     SUNLinSolFree(S);
@@ -149,7 +149,7 @@
 
 int SUNLinSolSetup_LapackDense(SUNLinearSolver S, SUNMatrix A)
 {
-  sunindextype n, ier;
+  int n, ier;
 
   /* check for valid inputs */
   if ((A == NULL) || (S == NULL)) { return SUN_ERR_ARG_CORRUPT; }
@@ -174,7 +174,7 @@
 int SUNLinSolSolve_LapackDense(SUNLinearSolver S, SUNMatrix A, N_Vector x,
                                N_Vector b, SUNDIALS_MAYBE_UNUSED sunrealtype tol)
 {
-  sunindextype n, one, ier;
+  int n, one, ier;
   sunrealtype* xdata;
 
   if ((A == NULL) || (S == NULL) || (x == NULL) || (b == NULL))
